### ISSUE TITLE
Save new credentials from Autofill

### ISSUE THREAD

**Issue by sashei** [mozilla-lockwise/lockwise-android#217]
### Acceptance Criteria

Save credentials when the user enters credentials that we don't have in Lockbox. For now, this will not respect app locked state; that will be handled in a follow-up ticket #632 or saving values from multi-screen login experiences #633 

### Engineering notes 

Handle the [`onSaveRequest`](https://developer.android.com/reference/android/service/autofill/AutofillService.html#onSaveRequest(android.service.autofill.SaveRequest,%20android.service.autofill.SaveCallback)) callback.

### Todo
- [x] add SaveInfo to the FillResponse
- [x] `onSuccess` called in `LockboxAutofillService` 
- [x] add username and pw autofill ids as parcelables to the client state in `onFillRequest` -- [documentation](https://developer.android.com/reference/android/service/autofill/AutofillService.html#saving-when-data-is-split-in-multiple-screens)
- [x] Modify `ParsedStructure` to implement `findNodeById`
- [x] Use `getAutofillValues` + map to `ServerPassword`
- [x] Implement `DataStoreAction.Add(credential: ServerPassword)`
- [x] Handle `DataStoreAction.Add` in `Datastore.kt`

**Comment by sashei** [comment]
I've done some more investigation, and I think we can avoid the `onSaveRequest` callback simply by not passing the [`SaveInfo`](https://developer.android.com/reference/android/service/autofill/SaveInfo) object to the FillResponse builder. I'll leave this story as a placeholder for further autofill work!!

**Comment by sandysage** [comment]
This is not a 'must have' for initial launch. It is something we need to provide prior to the Fenix launch to support our 'family of products' on Android

**Comment by eliserichards** [comment]
When saving new values, it's possible for both the username **and** password to be null. Currently, we will accept null usernames, but not null passwords. 

What error should we give to the user when they try to save invalid data (i.e. an empty password)? @changecourse @sandysage

**Comment by eliserichards** [comment]
Current save popup: 
<img width="350" alt="image" src="https://user-images.githubusercontent.com/43795363/58121211-b490cc80-7bbb-11e9-9e86-b5c16daa4efe.png">

**Comment by changecourse** [comment]
Per a conversation with @eliserichards, we have the following questions:

What are we able to display/have control of, from a UI perspective? Is it just the string, or do we have control over whether it is displayed as a toast vs a pop-over? 

Do we need a separate issue to handle "Success" use cases for saving a new login, e.g., a Confirmation dialog or toast message?

How are we handling de-duping for a use case involving the following... 
- I'm logging into DuoLingo with my credentials that are already present in Lockwise
- I choose to manually fill my login details instead of using Lockwise as an autofill provider
- Does Lockwise think this is a new entry? OR does it know to check, recognize that a matching entry already exists, and not display the "Add new login" prompt? 

How are we accounting for the use case where a password has changed for a given account, there is an old login on file, and we'd simply want to override/update the login with a new password, instead of creating a new entry entirely? 

I'd want to be sure that we define the acceptance criteria for all possible use cases as part of this flow and cover our bases for the alternate scenarios before shipping this as a one-off feature.

**Comment by changecourse** [comment]
@eliserichards looking back through this (for the first time since May) do you mind taking a stab at the questions I have above? That'll help me plan a design approach and unblock the issue. If not you, then perhaps @jhugman

**Comment by isabelrios** [comment]
Commenting as QA Needed label is added. 
I got this working using Douling app and latest master d04e5e3976b0715af02fa261bd53d5a31940d914. 
But triying with other apps, like Facebook and Twitter I could not get it working. 
More testing would be needed so leaving the label here and the issue under the QA column

**Comment by abodea** [comment]
I was also able to save a login using Duolingo and 9gag app.
I couldn't make it work using: Faceook, Facebook Lite, Instagram, Twitch, Amazon.
Note that Facebook got a save login feature integrated in the app, and maybe because of that, not any other app can save logins on their products.

**Comment by sandysage** [comment]
@abodea Unfortunately that has been the case for the autofill feature too. It's something we're considering for how we improve the app going forward: 1. defining some set of sites/apps where autofill is not working 2. triage/scope these cases and 3. come up with a plan for getting all of autofill (both fill and capture) working in more places. I'll create a ticket specifically for that initial definition work.

**Comment by abodea** [comment]
Verified as fixed on **v.4.0.0(6016)** using **Google Pixel 3 XL(Android 9), Samsung Galaxy S10+(Android 9).**
I verified it on the following: Spotify, Facebook, Twitter, Duolingo(App), Twitch, Yahoo, Pinterest.
